File: intellij/src/test/kotlin/ai/omnigent/intellij/conformance/Vectors.kt
package ai.omnigent.intellij.conformance

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import java.io.File

/**
 * Loads the language-neutral conformance vectors from docs/conformance/ (the
 * single source of truth SHARED with the TS suite — AC9). The Kotlin suite
 * loads these IDENTICALLY to the TS suite and must produce the same outputs.
 *
 * The conformance dir is located by walking up from the test working directory
 * until `docs/conformance` is found — works whether Gradle runs from the
 * intellij/ module dir or the repo root. Mirrors vscode/src/test/vectors.ts.
 */
object Vectors {
    val json = Json { ignoreUnknownKeys = true }

    val conformanceDir: File by lazy { findConformanceDir() }

    private fun findConformanceDir(): File {
        var dir: File? = File(System.getProperty("user.dir")).absoluteFile
        var i = 0
        while (dir != null && i < 8) {
            val candidate = File(dir, "docs/conformance")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile
            i++
        }
        error("could not locate docs/conformance from ${System.getProperty("user.dir")}")
    }

    fun load(name: String): JsonObject {
        val file = File(conformanceDir, name)
        return json.parseToJsonElement(file.readText()).let { it as JsonObject }
    }
}

File: vscode/src/test/vectors.ts
/**
 * Loads the language-neutral conformance vectors from docs/conformance/ (the
 * single source of truth shared with the future Kotlin impl). Tests read these
 * from the repo path; they contain no TS-isms.
 *
 * The conformance dir is located by walking up from the current working
 * directory until `docs/conformance` is found. This avoids `import.meta`
 * (incompatible with the CommonJS `tsc` graph) and works whether vitest runs
 * from the `vscode/` package dir or the repo root.
 */
import { readFileSync, existsSync } from "node:fs";
import { join, dirname } from "node:path";

function findConformanceDir(): string {
  let dir = process.cwd();
  for (let i = 0; i < 8; i++) {
    const candidate = join(dir, "docs", "conformance");
    if (existsSync(candidate)) {
      return candidate;
    }
    const parent = dirname(dir);
    if (parent === dir) {
      break;
    }
    dir = parent;
  }
  throw new Error("could not locate docs/conformance from " + process.cwd());
}

export const CONFORMANCE_DIR = findConformanceDir();

export function loadVectors<T = unknown>(name: string): T {
  const path = join(CONFORMANCE_DIR, name);
  return JSON.parse(readFileSync(path, "utf8")) as T;
}
